File: 129_B.py
n, m = map(int, input().split())
edges = []
for i in range(m):
    edges.append(list(map(int, input().split())))
nodes = [i + 1 for i in range(n)]
count = 0
while True:
    tied = {}
    for node in nodes:
        tied[node] = 0
    for edge in edges:
        tied[edge[0]] += 1
        tied[edge[1]] += 1
    isolated = [node for node in tied.keys() if tied[node] == 1]
    if not isolated:
        break
    nodes = [node for node in nodes if node not in isolated]
    edges = [edge for edge in edges if edge[0] in nodes and edge[1] in nodes]
    count += 1
print(count)

File: 129B.cpp
n, m = map(int, input().split())
e = [list(map(int, input().split())) for _ in range(m)]
v = set(range(1, n + 1))
c = 0

while True:
    t = {node: 0 for node in v}
    for edge in e:
        t[edge[0]] += 1
        t[edge[1]] += 1
    
    s = {node for node, degree in t.items() if degree == 1}
    if not s:
        break
    
    v -= s
    e = [edge for edge in e if edge[0] in v and edge[1] in v]
    c += 1

print(c)
